util/include/channels/mpsc_bounded_channel.h

```diff
@@ -0,0 +1,226 @@
+  }
+
+  std::optional<size_t> size() { return queue_->size(); }
+  std::optional<size_t> capacity() { return queue_->capacity(); }
+
+  ~BoundedMpscSender() {
+    if (!queue_) return;
+    queue_->removeSender();
+  }
+
+  BoundedMpscSender(const BoundedMpscSender& other) : BoundedMpscSender(other.queue_) {}
+  BoundedMpscSender& operator=(const BoundedMpscSender& other) {
+    *this = BoundedMpscSender(other);
+    return *this;
+  }
+  BoundedMpscSender(BoundedMpscSender&&) noexcept = default;
+  BoundedMpscSender& operator=(BoundedMpscSender&&) noexcept = default;
+
+ private:
+  BoundedMpscSender(const std::shared_ptr<impl::BoundedMpscQueue<Msg>>& queue) : queue_(queue) { queue_->addSender(); }
+
+  std::shared_ptr<impl::BoundedMpscQueue<Msg>> queue_;
+
+  template <typename T>
+  friend std::pair<BoundedMpscSender<T>, BoundedMpscReceiver<T>> makeBoundedMpscChannel(size_t capacity);
+};  // namespace concord::channel
+
+template <typename Msg>
+class BoundedMpscReceiver {
+ public:
+  // Receive a message over a channel.
+  // Block waiting indefinitely.
+  //
+  // Throws NoSendersError if all senders have been destructed and there are no more messages left
+  // in the communication buffer.
+  Msg recv() {
+    std::unique_lock<std::mutex> lock(queue_->mutex_);
+    queue_->cv_.wait(lock, [this]() { return waitPred(); });
+
+    // Always return a value if it's available, even if there are no senders.
+    if (!queue_->buf_.empty()) {
+      return pop();
+    }
+    throw NoSendersError();
+  }
+
+  // Receive a message over a channel and return it.
+  // Return std::nullopt on timeout.
+  //
+  // Throws NoSendersError if all senders have been destructed and there are no more messages left
+  // in the communication buffer.
+  std::optional<Msg> recv(std::chrono::milliseconds timeout) {
+    std::unique_lock<std::mutex> lock(queue_->mutex_);
+    if (!queue_->cv_.wait_for(lock, timeout, [this]() { return waitPred(); })) {
+      // Timeout
+      return std::nullopt;
+    }
+
+    // Always return a value if it's available, even if there are no senders.
+    if (!queue_->buf_.empty()) {
+      return pop();
+    }
+    throw NoSendersError();
+  }
+
+  std::optional<size_t> size() { return queue_->size(); }
+  std::optional<size_t> capacity() { return queue_->capacity(); }
+
+  ~BoundedMpscReceiver() {
+    if (queue_) {
+      queue_->invalidateReceiver();
+    }
+  }
+
+  // You can't copy an MPSC receiver
+  BoundedMpscReceiver(const BoundedMpscReceiver&) = delete;
+  BoundedMpscReceiver& operator=(const BoundedMpscReceiver&) = delete;
+
+  // We do allow moving a receiver though
+  BoundedMpscReceiver(BoundedMpscReceiver&&) noexcept = default;
+  BoundedMpscReceiver& operator=(BoundedMpscReceiver&&) noexcept = default;
+
+ private:
+  BoundedMpscReceiver(const std::shared_ptr<impl::BoundedMpscQueue<Msg>>& queue) : queue_(queue) {}
+
+  Msg pop() {
+    auto msg = std::move(queue_->buf_.front());
+    queue_->buf_.pop();
```

Comment:
Nit: If we ever use a buffer type for which `pop()` can throw, then that would be exception-unsafe as we've already moved. Not sure if this is worth documenting.

Reply:
In this case, we use a std::deque, and pop cannot throw since we always call this on a non-empty deque. I will add a comment.